app/src/main/java/org/blitzortung/android/data/provider/blitzortung/BlitzortungHttpDataProvider.kt

```diff
@@ -135,35 +111,39 @@ class BlitzortungHttpDataProvider @JvmOverloads constructor(private val urlForma
         return reader
     }
 
-    override fun getStations(region: Int): List<Station> {
-        val stations = ArrayList<Station>()
+    /**
+     * Used to retrieve Data from Blitzortung
+     * @param readerSeq A sequence of nullable BufferedReader, which the data is read from
+     * @param parse A Lambda which receives a sequence of lines from a buffered reader and transforms them into a sequence of T
+     * @return Returns a list of parsed T's
+     */
+    private fun <T: Any> retrieveData(logMessage: String, readerSeq: Sequence<BufferedReader?>, parse: (String) -> T?): List<T> {
+        var size = 0
 
         val username = username
         val password = password
-        if (username != null && password != null) {
-
-            val reader = readFromUrl(Type.STATIONS, region)
-
-            var size = 0
-            reader?.use { reader ->
-                reader.forEachLine { line ->
-                    size += line.length
-                    try {
-                        val station = stationMapBuilder.buildFromLine(line)
-                        stations.add(station)
-                    } catch (e: NumberFormatException) {
-                        Log.w(Main.LOG_TAG, "BlitzortungHttpProvider: error parsing '%s'".format(line))
-                    }
-                }
+        if (username == null || password == null)
+            throw RuntimeException("no credentials provided")
+
+        return readerSeq.filterNotNull().flatMap {
+            val tmpList = it.lineSequence().mapNotNull {
+                size += it.length
+
+                return@mapNotNull parse(it)
             }
+
             Log.v(Main.LOG_TAG,
-                    "BlitzortungHttpProvider: read %d bytes (%d stations) from region %d".format(size, stations.size, region))
+                    logMessage.format(size, tmpList.count()))
```

Comment:
Calling count on the sequence led to an error because of an already consumed sequence when putting them together in the `.toList()`call later. This was fixed with 0ff98d17ad073eb71d29d7502aecd1b191791cab.